api/grpcserver/mesh_service.go

```diff
@@ -111,7 +114,140 @@ func (s MeshService) MaxTransactionsPerSecond(ctx context.Context, in *pb.MaxTra
 // AccountMeshDataQuery returns account data
 func (s MeshService) AccountMeshDataQuery(ctx context.Context, in *pb.AccountMeshDataQueryRequest) (*pb.AccountMeshDataQueryResponse, error) {
 	log.Info("GRPC MeshService.AccountMeshDataQuery")
-	return nil, nil
+
+	// This is the latest layer we're aware of
+	//latestLayer := s.Mesh.LatestLayer()
+	startLayer := types.LayerID(in.MinLayer)
+
+	if startLayer > s.Mesh.LatestLayer() {
+		return nil, status.Errorf(codes.InvalidArgument, "`LatestLayer` must be less than or equal to latest layer")
+	}
+	if in.Filter == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "`Filter` must be provided")
+	}
+	if in.Filter.AccountId == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "`Filter.AccountId` must be provided")
+	}
+
+	// Gather transaction data
+	//txs := &pb.AccountMeshData{DataItem: &pb.AccountMeshData_Transaction{Transaction: }
+	addr := types.BytesToAddress(in.Filter.AccountId.Address)
+	res := &pb.AccountMeshDataQueryResponse{}
+	if in.Filter.AccountMeshDataFlags&uint32(pb.AccountMeshDataFlag_ACCOUNT_MESH_DATA_FLAG_TRANSACTIONS) != 0 {
+		meshTxIds := s.getTxIdsFromMesh(startLayer, addr)
+		mempoolTxIds := s.Mempool.GetTxIdsByAddress(addr)
+
+		// Look up full data for all unique txids
+		txs, missing := s.Mesh.GetTransactions(append(meshTxIds, mempoolTxIds...))
+
+		// TODO: Do we ever expect txs to be missing here?
+		// E.g., if this node has not synced/received them yet.
+		if len(missing) != 0 {
+			log.Error("could not find transactions %v", missing)
+			return nil, status.Errorf(codes.Internal, "error retrieving tx data")
+		}
+		for _, t := range txs {
+			res.Data = append(res.Data, &pb.AccountMeshData{
+				DataItem: &pb.AccountMeshData_Transaction{
+					Transaction: convertTransaction(t),
+				},
+			})
+		}
+	}
+
+	// Gather activation data
+	if in.Filter.AccountMeshDataFlags&uint32(pb.AccountMeshDataFlag_ACCOUNT_MESH_DATA_FLAG_ACTIVATIONS) != 0 {
+		// TODO: we have no way to look up activations by coinbase
+		// so we have no choice but to read all of them (!!!)
+		var activations []types.ATXID
+		for l := startLayer; l <= s.Mesh.LatestLayer(); l++ {
+			layer, err := s.Mesh.GetLayer(l)
```

Comment:
we should index the atx ids by layer, GetLayer is very costly and takes time. there is a `GetEpochAtxs` method (perhaps only in my pr, but anyways, if we need to query for this data it needs to be better indexed in the code

Reply:
I agree but that task seems somewhat orthogonal to this implementation work - I consider that an optimization that we can add later. Do you think we can live with this in the short term, and prioritize adding indexes after the basic implementation is done?

Reply:
Opened #2064 for this

Reply:
I would prefer if we index the data before making this code operational and letting others use it

Reply:
My approach so far has been, implement the API in the simplest, most naive way possible, with the fewest changes to existing code, to begin, then we can iterate and optimize it. Maybe let's discuss next week with @avive.